cpp/src/parquet/encoding.cc

```diff
@@ -116,23 +109,22 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
 
   void Put(const ByteArray& val) {
     // Write the result to the output stream
-    PARQUET_THROW_NOT_OK(values_sink_->Write(reinterpret_cast<const uint8_t*>(&val.len),
-                                             sizeof(uint32_t)));
-    if (val.len > 0) {
-      DCHECK(nullptr != val.ptr) << "Value ptr cannot be NULL";
+    const int64_t increment = static_cast<int64_t>(val.len + sizeof(uint32_t));
+    if (ARROW_PREDICT_FALSE(sink_.length() + increment > sink_.capacity())) {
+      PARQUET_THROW_NOT_OK(sink_.Reserve(increment));
     }
-    PARQUET_THROW_NOT_OK(
-        values_sink_->Write(reinterpret_cast<const uint8_t*>(val.ptr), val.len));
+    DCHECK(val.len == 0 || nullptr != val.ptr) << "Value ptr cannot be NULL";
+    sink_.UnsafeAppend(&val.len, sizeof(uint32_t));
+    sink_.UnsafeAppend(val.ptr, val.len);
   }
 
  protected:
-  std::shared_ptr<arrow::io::BufferOutputStream> values_sink_;
+  arrow::BufferBuilder sink_;
 };
 
 template <typename DType>
 void PlainEncoder<DType>::Put(const T* buffer, int num_values) {
```

Comment:
It looks like Travis is failing because BufferBuilder::Append ultimately uses memcpy and memset (without checking whether the length of data appended is 0), and they have the `__nonnull` attribute set for their pointer arguments. A check for length == 0 should be added either here or in BufferBuilder